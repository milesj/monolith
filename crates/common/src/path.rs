pub use relative_path::*;
use starbase_styles::color;
use std::path::Path;

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[cfg(unix)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

pub enum RelativeFrom<'path> {
    Project(&'path str),
    Workspace,
}

pub fn expand_to_workspace_relative<P: AsRef<str>>(
    from_format: RelativeFrom,
    path: P,
) -> WorkspaceRelativePathBuf {
    let path = standardize_separators(path.as_ref());

    match from_format {
        RelativeFrom::Project(source) => {
            // Root-level project
            if source.is_empty() || source == "." {
                WorkspaceRelativePathBuf::from(path)

                // Project-level, prefix with source path
            } else {
                let project_source = standardize_separators(source);

                if let Some(negated_glob) = path.strip_prefix('!') {
                    WorkspaceRelativePathBuf::from(format!("!{project_source}")).join(negated_glob)
                } else {
                    WorkspaceRelativePathBuf::from(project_source).join(path)
                }
            }
        }
        RelativeFrom::Workspace => WorkspaceRelativePathBuf::from(path),
    }
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let path = path.as_ref();

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            color::path(path)
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    #[cfg(windows)]
    {
        format!("{}.exe", name.as_ref())
    }

    #[cfg(not(windows))]
    {
        name.as_ref().into()
    }
}
