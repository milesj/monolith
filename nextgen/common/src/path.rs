pub use relative_path::{RelativePath, RelativePathBuf};

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}
