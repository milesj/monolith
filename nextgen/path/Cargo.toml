[package]
name = "moon_path"
version = "0.1.0"
edition = "2021"

[dependencies]
relative-path = { version = "1.8.0", features = ["serde"] }
serde = { workspace = true }
