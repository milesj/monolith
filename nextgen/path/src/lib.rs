pub use relative_path::{RelativePath, RelativePathBuf};

// pub enum PathType {
//     File(String),
//     Glob(String),
// }

// pub enum Location<T> {
//     Absolute(T),
//     ProjectRelative(T),
//     WorkspaceRelative(T),
// }

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[inline]
pub fn expand_to_workspace_relative<F, P>(file: F, project_source: P) -> WorkspaceRelativePathBuf
where
    F: AsRef<str>,
    P: AsRef<str>,
{
    let file = file.as_ref();
    let project_source = project_source.as_ref();

    // Workspace relative negative glob
    if file.starts_with("/!") || file.starts_with("!/") {
        return WorkspaceRelativePathBuf::from(format!("!{}", standardize_separators(&file[2..])))
            .normalize();
    }

    // Workspace relative file/glob
    if let Some(path) = file.strip_prefix('/') {
        return WorkspaceRelativePathBuf::from(standardize_separators(path)).normalize();
    }

    // Project relative negative glob
    if let Some(glob) = file.strip_prefix('!') {
        return WorkspaceRelativePathBuf::from(format!("!{}", project_source))
            .join(standardize_separators(glob))
            .normalize();
    }

    // Project relative file/glob
    WorkspaceRelativePathBuf::from(project_source)
        .join(standardize_separators(file))
        .normalize()
}
