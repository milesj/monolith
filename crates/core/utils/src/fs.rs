use async_recursion::async_recursion;
use futures::future::try_join_all;
use json_comments::StripComments;
use moon_error::{map_io_to_fs_error, map_json_to_error, map_yaml_to_error, MoonError};
use regex::Regex;
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime};
use tokio::fs;

#[inline]
pub fn clean_json<T: AsRef<str>>(json: T) -> Result<String, MoonError> {
    let json = json.as_ref();

    // Remove comments
    let mut stripped = String::with_capacity(json.len());

    StripComments::new(json.as_bytes())
        .read_to_string(&mut stripped)
        .map_err(MoonError::Unknown)?;

    // Remove trailing commas
    let stripped = Regex::new(r",(?P<valid>\s*})")
        .unwrap()
        .replace_all(&stripped, "$valid");

    Ok(String::from(stripped))
}

#[inline]
pub async fn copy_file<S: AsRef<Path>, D: AsRef<Path>>(from: S, to: D) -> Result<(), MoonError> {
    let from = from.as_ref();
    let to = to.as_ref();
    let to_dir = to.parent().unwrap();

    create_dir_all(to_dir).await?;

    fs::copy(from, to)
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[async_recursion]
pub async fn copy_dir_all<T: AsRef<Path> + Send>(
    from_root: T,
    from: T,
    to_root: T,
) -> Result<(), MoonError> {
    let from_root = from_root.as_ref();
    let from = from.as_ref();
    let to_root = to_root.as_ref();
    let entries = read_dir(from).await?;
    let mut files = vec![];
    let mut dirs = vec![];

    for entry in entries {
        let path = entry.path();

        if path.is_file() {
            files.push(copy_file(
                path.to_owned(),
                to_root.join(path.strip_prefix(from_root).unwrap()),
            ));
        } else {
            dirs.push(path);
        }
    }

    // Copy files before dirs incase an error occurs
    try_join_all(files).await?;

    // Copy dirs in sequence for the same reason
    for dir in dirs {
        copy_dir_all(from_root, &dir, to_root).await?;
    }

    Ok(())
}

#[inline]
pub async fn create_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if !path.exists() {
        fs::create_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub fn find_upwards<F, P>(name: F, dir: P) -> Option<PathBuf>
where
    F: AsRef<str>,
    P: AsRef<Path>,
{
    let dir = dir.as_ref();
    let findable = dir.join(name.as_ref());

    if findable.exists() {
        return Some(findable);
    }

    match dir.parent() {
        Some(parent_dir) => find_upwards(name, parent_dir),
        None => None,
    }
}

#[inline]
pub async fn metadata<T: AsRef<Path>>(path: T) -> Result<std::fs::Metadata, MoonError> {
    let path = path.as_ref();

    fs::metadata(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))
}

#[inline]
pub async fn read_dir<T: AsRef<Path>>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        results.push(entry);
    }

    Ok(results)
}

#[async_recursion]
pub async fn read_dir_all<T: AsRef<Path> + Send>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        if let Ok(file_type) = entry.file_type().await {
            if file_type.is_dir() {
                results.extend(read_dir_all(&entry.path()).await?);
            } else {
                results.push(entry);
            }
        }
    }

    Ok(results)
}

#[inline]
pub async fn read<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    let path = path.as_ref();
    let data = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(data)
}

#[inline]
pub async fn read_json<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = read_json_string(path).await?;

    let json: D =
        serde_json::from_str(&contents).map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    clean_json(read(path).await?)
}

#[inline]
pub async fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    let json: D =
        serde_yaml::from_str(&contents).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn remove<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.is_file() {
        remove_file(path).await?;
    } else if path.is_dir() {
        remove_dir_all(path).await?;
    }

    Ok(())
}

#[inline]
pub async fn remove_file<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_file(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub async fn remove_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

pub type RemoveDirContentsResult = (usize, u64);

pub async fn remove_dir_stale_contents<P: AsRef<Path>>(
    dir: P,
    duration: Duration,
) -> Result<RemoveDirContentsResult, MoonError> {
    let mut files_deleted: usize = 0;
    let mut bytes_saved: u64 = 0;
    let threshold = SystemTime::now() - duration;

    for entry in read_dir(dir.as_ref()).await? {
        let path = entry.path();

        if path.is_file() {
            let mut bytes = 0;

            if let Ok(metadata) = entry.metadata().await {
                bytes = metadata.len();

                if let Ok(filetime) = metadata.accessed().or_else(|_| metadata.created()) {
                    if filetime > threshold {
                        // Not stale yet
                        continue;
                    }
                } else {
                    // Not supported in environment
                    continue;
                }
            }

            if remove_file(path).await.is_ok() {
                files_deleted += 1;
                bytes_saved += bytes;
            }
        }
    }

    Ok((files_deleted, bytes_saved))
}

#[inline]
pub async fn rename<F: AsRef<Path>, T: AsRef<Path>>(from: F, to: T) -> Result<(), MoonError> {
    let from = from.as_ref();

    fs::rename(from, to.as_ref())
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write<T: AsRef<Path>>(path: T, data: impl AsRef<[u8]>) -> Result<(), MoonError> {
    let path = path.as_ref();

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_json<P, D>(path: P, json: &D, pretty: bool) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data = if pretty {
        serde_json::to_string_pretty(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    } else {
        serde_json::to_string(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    };

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_yaml<P, D>(path: P, yaml: &D) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data =
        serde_yaml::to_string(&yaml).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

pub mod temp {
    use super::*;
    use moon_constants::CONFIG_DIRNAME;
    use std::fs; // TEMPORARILY sync

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache/temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{:x}.{}", md5::compute(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> Result<Option<String>, MoonError> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        // Temp files only last for 4 hours (half a workday)
        let threshold = SystemTime::now() - Duration::from_secs(60 * 60 * 4);

        if let Ok(metadata) = file.metadata() {
            if let Ok(filetime) = metadata.created() {
                if filetime > threshold {
                    fs::remove_file(file)?;

                    return Ok(None);
                }
            }
        }

        Ok(Some(fs::read_to_string(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> Result<(), MoonError> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write(path, data.as_ref())?;

        Ok(())
    }
}

pub mod sync {
    use super::*;
    use std::fs::read_to_string;

    #[inline]
    pub fn read<P>(path: P) -> Result<String, MoonError>
    where
        P: AsRef<Path>,
    {
        let path = path.as_ref();
        let data = read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        Ok(data)
    }

    #[inline]
    pub fn read_json<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents = read_json_string(path)?;

        let json: D = serde_json::from_str(&contents)
            .map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }

    #[inline]
    pub fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
        clean_json(read(path.as_ref())?)
    }

    #[inline]
    pub fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents =
            read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        let json: D = serde_yaml::from_str(&contents)
            .map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }
}
