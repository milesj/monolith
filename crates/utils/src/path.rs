pub use dirs::home_dir as get_home_dir;
use moon_error::MoonError;
use path_clean::PathClean;
use std::path::{Path, PathBuf};

/// If a file starts with "/", expand from the workspace root, otherwise the project root.
pub fn expand_root_path(file: &str, workspace_root: &Path, project_root: &Path) -> PathBuf {
    if file.starts_with('/') {
        workspace_root.join(file.strip_prefix('/').unwrap())
    } else {
        project_root.join(file)
    }
}

// This is not very exhaustive and may be inaccurate.
pub fn is_glob(value: &str) -> bool {
    let single_values = vec!['*', '?', '1'];
    let paired_values = vec![('{', '}'), ('[', ']')];
    let mut bytes = value.bytes();
    let mut is_escaped = |index: usize| {
        if index == 0 {
            return false;
        }

        bytes.nth(index - 1).unwrap_or(b' ') == b'\\'
    };

    if value.contains("**") {
        return true;
    }

    for single in single_values {
        if !value.contains(single) {
            continue;
        }

        if let Some(index) = value.find(single) {
            if !is_escaped(index) {
                return true;
            }
        }
    }

    for (open, close) in paired_values {
        if !value.contains(open) || !value.contains(close) {
            continue;
        }

        if let Some(index) = value.find(open) {
            if !is_escaped(index) {
                return true;
            }
        }
    }

    false
}

pub fn is_path_glob(path: &Path) -> bool {
    is_glob(&path.to_string_lossy())
}

pub fn normalize(path: &Path) -> PathBuf {
    path.to_path_buf().clean()
}

pub fn normalize_glob(path: &Path) -> String {
    // Always use forward slashes for globs
    let glob = standardize_separators(&path.to_string_lossy());

    // Remove UNC prefix as it breaks glob matching
    if std::env::consts::OS == "windows" {
        return glob.replace("//?/", "");
    }

    glob
}

#[cfg(not(windows))]
pub fn normalize_separators(path: &str) -> String {
    path.replace('\\', "/")
}

#[cfg(windows)]
pub fn normalize_separators(path: &str) -> String {
    path.replace('/', "\\")
}

pub fn path_to_string(path: &Path) -> Result<String, MoonError> {
    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(MoonError::PathInvalidUTF8(path.to_path_buf())),
    }
}

pub fn replace_home_dir(value: &str) -> String {
    if let Some(home_dir) = get_home_dir() {
        let home_dir_str = home_dir.to_str().unwrap();

        // Replace both forward and backward slashes
        return value
            .replace(home_dir_str, "~")
            .replace(&standardize_separators(home_dir_str), "~");
    }

    value.to_owned()
}

pub fn standardize_separators(path: &str) -> String {
    path.replace('\\', "/")
}

#[cfg(test)]
mod tests {
    use super::*;

    mod is_glob {
        use super::*;

        #[test]
        fn returns_true_when_a_glob() {
            assert!(is_glob("**"));
            assert!(is_glob("**/src/*"));
            assert!(is_glob("src/**"));
            assert!(is_glob("*.ts"));
            assert!(is_glob("file.*"));
            assert!(is_glob("file.{js,ts}"));
            assert!(is_glob("file.[jstx]"));
            assert!(is_glob("file.tsx?"));
        }

        #[test]
        fn returns_false_when_not_glob() {
            assert!(!is_glob("dir"));
            assert!(!is_glob("file.rs"));
            assert!(!is_glob("dir/file.ts"));
            assert!(!is_glob("dir/dir/file_test.rs"));
            assert!(!is_glob("dir/dirDir/file-ts.js"));
        }

        #[test]
        fn returns_false_when_escaped_glob() {
            assert!(!is_glob("\\*.rs"));
            assert!(!is_glob("file\\?.js"));
            assert!(!is_glob("folder-\\[id\\]"));
        }
    }
}
