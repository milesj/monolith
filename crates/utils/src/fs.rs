use async_recursion::async_recursion;
use json_comments::StripComments;
use moon_error::{map_io_to_fs_error, map_json_to_error, MoonError};
use regex::Regex;
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::io::Read;
use std::path::{Path, PathBuf};
use tokio::fs;

pub use dirs::home_dir as get_home_dir;

pub fn clean_json(json: String) -> Result<String, MoonError> {
    // Remove comments
    let mut stripped = String::with_capacity(json.len());

    StripComments::new(json.as_bytes())
        .read_to_string(&mut stripped)
        .map_err(MoonError::Unknown)?;

    // Remove trailing commas
    let stripped = Regex::new(r",(?P<valid>\s*})")
        .unwrap()
        .replace_all(&stripped, "$valid");

    Ok(String::from(stripped))
}

pub async fn create_dir_all(path: &Path) -> Result<(), MoonError> {
    if !path.exists() {
        fs::create_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

/// If a file starts with "/", expand from the workspace root, otherwise the project root.
pub fn expand_root_path(file: &str, workspace_root: &Path, project_root: &Path) -> PathBuf {
    if file.starts_with('/') {
        workspace_root.join(file.strip_prefix('/').unwrap())
    } else {
        project_root.join(file)
    }
}

// This is not very exhaustive and may be inaccurate.
pub fn is_glob(value: &str) -> bool {
    let single_values = vec!['*', '?', '1'];
    let paired_values = vec![('{', '}'), ('[', ']')];
    let mut bytes = value.bytes();
    let mut is_escaped = |index: usize| {
        if index == 0 {
            return false;
        }

        bytes.nth(index - 1).unwrap_or(b' ') == b'\\'
    };

    if value.contains("**") {
        return true;
    }

    for single in single_values {
        if !value.contains(single) {
            continue;
        }

        if let Some(index) = value.find(single) {
            if !is_escaped(index) {
                return true;
            }
        }
    }

    for (open, close) in paired_values {
        if !value.contains(open) || !value.contains(close) {
            continue;
        }

        if let Some(index) = value.find(open) {
            if !is_escaped(index) {
                return true;
            }
        }
    }

    false
}

pub fn is_path_glob(path: &Path) -> bool {
    is_glob(&path.to_string_lossy())
}

pub async fn link_file(from_root: &Path, from: &Path, to_root: &Path) -> Result<(), MoonError> {
    let to = to_root.join(from.strip_prefix(from_root).unwrap());

    // Hardlink has already been created
    if to.exists() {
        return Ok(());
    }

    let to_dir = to.parent().unwrap();

    if to_dir != to_root {
        create_dir_all(to_dir).await?;
    }

    fs::hard_link(from, &to)
        .await
        .map_err(|_| MoonError::HardLink(from.to_path_buf(), to.clone()))?;

    Ok(())
}

#[async_recursion]
pub async fn link_dir(from_root: &Path, from: &Path, to_root: &Path) -> Result<(), MoonError> {
    let entries = read_dir(from).await?;
    let mut dirs = vec![];

    // Link files before dirs incase an error occurs
    for entry in entries {
        let path = entry.path();

        if path.is_file() {
            link_file(from_root, &path, to_root).await?;
        } else {
            dirs.push(path);
        }
    }

    // Link dirs in sequence for the same reason
    for dir in dirs {
        link_dir(from_root, &dir, to_root).await?;
    }

    Ok(())
}

pub async fn metadata(path: &Path) -> Result<std::fs::Metadata, MoonError> {
    Ok(fs::metadata(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?)
}

pub fn normalize_glob(path: &Path) -> String {
    // Always use forward slashes for globs
    let glob = path.to_string_lossy().replace("\\", "/");

    if std::env::consts::OS == "windows" {
        return glob.replace("//?/", ""); // Is this needed for globs?
    }

    glob
}

pub async fn read_dir(path: &Path) -> Result<Vec<fs::DirEntry>, MoonError> {
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        results.push(entry);
    }

    Ok(results)
}

pub async fn read_json<T>(path: &Path) -> Result<T, MoonError>
where
    T: DeserializeOwned,
{
    let contents = read_json_string(path).await?;

    let json: T =
        serde_json::from_str(&contents).map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

pub async fn read_json_string(path: &Path) -> Result<String, MoonError> {
    let json = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(clean_json(json)?)
}

pub async fn remove_file(path: &Path) -> Result<(), MoonError> {
    if path.exists() {
        fs::remove_file(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

pub async fn remove_dir_all(path: &Path) -> Result<(), MoonError> {
    if path.exists() {
        fs::remove_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

pub async fn write(path: &Path, data: impl AsRef<[u8]>) -> Result<(), MoonError> {
    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

pub async fn write_json<T>(path: &Path, json: &T, pretty: bool) -> Result<(), MoonError>
where
    T: ?Sized + Serialize,
{
    let data = if pretty {
        serde_json::to_string_pretty(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    } else {
        serde_json::to_string(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    };

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    mod is_glob {
        use super::*;

        #[test]
        fn returns_true_when_a_glob() {
            assert!(is_glob("**"));
            assert!(is_glob("**/src/*"));
            assert!(is_glob("src/**"));
            assert!(is_glob("*.ts"));
            assert!(is_glob("file.*"));
            assert!(is_glob("file.{js,ts}"));
            assert!(is_glob("file.[jstx]"));
            assert!(is_glob("file.tsx?"));
        }

        #[test]
        fn returns_false_when_not_glob() {
            assert!(!is_glob("dir"));
            assert!(!is_glob("file.rs"));
            assert!(!is_glob("dir/file.ts"));
            assert!(!is_glob("dir/dir/file_test.rs"));
            assert!(!is_glob("dir/dirDir/file-ts.js"));
        }

        #[test]
        fn returns_false_when_escaped_glob() {
            assert!(!is_glob("\\*.rs"));
            assert!(!is_glob("file\\?.js"));
            assert!(!is_glob("folder-\\[id\\]"));
        }
    }
}
