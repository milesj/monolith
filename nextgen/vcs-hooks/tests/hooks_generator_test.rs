use moon_config::VcsConfig;
use moon_vcs::{BoxedVcs, Git};
use moon_vcs_hooks::HooksGenerator;
use rustc_hash::FxHashMap;
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use std::fs;
use std::path::Path;

fn load_git(root: &Path) -> BoxedVcs {
    Box::new(Git::load(root, "master", &[]).unwrap())
}

async fn run_generator(root: &Path) {
    HooksGenerator::new(
        root,
        &load_git(root),
        &VcsConfig {
            hooks: FxHashMap::from_iter([
                (
                    "pre-commit".into(),
                    vec!["moon run :lint".into(), "some-command".into()],
                ),
                ("post-push".into(), vec!["moon check --all".into()]),
            ]),
            ..VcsConfig::default()
        },
    )
    .generate()
    .await
    .unwrap();
}

#[tokio::test]
async fn doesnt_generate_when_no_hooks() {
    let sandbox = create_empty_sandbox();
    sandbox.enable_git();

    HooksGenerator::new(
        sandbox.path(),
        &load_git(sandbox.path()),
        &VcsConfig::default(),
    )
    .generate()
    .await
    .unwrap();

    assert!(!sandbox.path().join(".moon/hooks").exists());
}

#[tokio::test]
async fn doesnt_generate_when_no_commands() {
    let sandbox = create_empty_sandbox();
    sandbox.enable_git();

    HooksGenerator::new(
        sandbox.path(),
        &load_git(sandbox.path()),
        &VcsConfig {
            hooks: FxHashMap::from_iter([
                ("pre-commit".into(), vec![]),
                ("post-push".into(), vec![]),
            ]),
            ..VcsConfig::default()
        },
    )
    .generate()
    .await
    .unwrap();

    assert!(!sandbox.path().join(".moon/hooks").exists());
}

#[cfg(not(windows))]
mod unix {
    use super::*;

    #[tokio::test]
    async fn creates_local_hook_files() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.sh");
        let post_push = sandbox.path().join(".moon/hooks/post-push.sh");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());
    }

    #[tokio::test]
    async fn links_git_hooks() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());
    }
}

#[cfg(windows)]
mod windows {
    use super::*;

    #[tokio::test]
    async fn creates_local_hook_files() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.ps1");
        let post_push = sandbox.path().join(".moon/hooks/post-push.ps1");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());
    }

    #[tokio::test]
    async fn links_git_hooks() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());
    }
}
