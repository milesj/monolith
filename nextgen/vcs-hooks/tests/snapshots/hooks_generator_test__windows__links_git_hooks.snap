---
source: nextgen/vcs-hooks/tests/hooks_generator_test.rs
expression: "out1.replace(\"powershell.exe\", \"pwsh.exe\")"
---
#!/bin/sh
pwsh.exe -NoProfile -ExecutionPolicy Bypass -File '.\.moon\hooks\pre-commit.ps1'
